tests/metrics/estimate_gas.py

```diff
@@ -118,6 +119,76 @@ def start_collection(self) -> None:
+            .format(compressed=False)
+        if pubkey_bytes != umbral_pubkey_bytes:
+            recoverable_signature = bytes(signature) + bytes([1])
+        return recoverable_signature
+
+    delegating_privkey = UmbralPrivateKey.gen_key()
+    _symmetric_key, capsule = pre._encapsulate(delegating_privkey.get_pubkey())
+    signing_privkey = UmbralPrivateKey.gen_key()
+    signer = Signer(signing_privkey)
+    priv_key_bob = UmbralPrivateKey.gen_key()
+    pub_key_bob = priv_key_bob.get_pubkey()
+    kfrags = pre.generate_kfrags(delegating_privkey=delegating_privkey,
+                                 signer=signer,
+                                 receiving_pubkey=pub_key_bob,
+                                 threshold=2,
+                                 N=4,
+                                 sign_delegating_key=False,
+                                 sign_receiving_key=False)
+    capsule.set_correctness_keys(delegating_privkey.get_pubkey(), pub_key_bob, signing_privkey.get_pubkey())
+    cfrag = pre.reencrypt(kfrags[0], capsule, metadata=os.urandom(34))
+    capsule_bytes = capsule.to_bytes()
+    if corrupt:
+        cfrag.proof.bn_sig = CurveBN.gen_rand(capsule.params.curve)
+    cfrag_bytes = cfrag.to_bytes()
+    hash_ctx = hashes.Hash(hashes.SHA256(), backend=backend)
+    hash_ctx.update(capsule_bytes + cfrag_bytes)
+    requester_umbral_private_key = UmbralPrivateKey.gen_key()
+    requester_umbral_public_key_bytes = requester_umbral_private_key.get_pubkey().to_bytes(is_compressed=False)
+    capsule_signature_by_requester = sign_data(capsule_bytes, requester_umbral_private_key)
+    miner_umbral_private_key = UmbralPrivateKey.gen_key()
+    miner_umbral_public_key_bytes = miner_umbral_private_key.get_pubkey().to_bytes(is_compressed=False)
+    # Sign Umbral public key using eth-key
+    hash_ctx = hashes.Hash(hashes.SHA256(), backend=backend)
+    hash_ctx.update(miner_umbral_public_key_bytes)
+    miner_umbral_public_key_hash = hash_ctx.finalize()
+    address = to_canonical_address(miner)
+    sig_key = testerchain.interface.provider.ethereum_tester.backend._key_lookup[address]
+    signed_miner_umbral_public_key = bytes(sig_key.sign_msg_hash(miner_umbral_public_key_hash))
+
+    capsule_signature_by_requester_and_miner = sign_data(capsule_signature_by_requester, miner_umbral_private_key)
+    cfrag_signature_by_miner = sign_data(cfrag_bytes, miner_umbral_private_key)
+    evidence = IndisputableEvidence(capsule, cfrag, ursula=None)
+    evidence_data = evidence.precompute_values()
+    return (capsule_bytes,
+            capsule_signature_by_requester,
+            capsule_signature_by_requester_and_miner,
+            cfrag_bytes,
+            cfrag_signature_by_miner,
+            requester_umbral_public_key_bytes,
+            miner_umbral_public_key_bytes,
+            signed_miner_umbral_public_key,
+            evidence_data)
+
+
```

Comment:
These logic seems very similar to the one in the test for MiningAdjudicator and the test for intercontract integration. Perhaps you could reuse this somehow.

Reply:
Yeah, exactly. @KPrasch how better to organize this methods? Maybe some of them will be in the main code?